app/modules/entities/components/layouts/works_browser_section.svelte

```diff
@@ -1,50 +1,131 @@
   import { bySearchMatchScore, getSelectedUris } from '#entities/components/lib/works_browser_helpers'
-  import { onChange } from '#lib/svelte/svelte'
   import { flip } from 'svelte/animate'
+  import { i18n } from '#user/lib/i18n'
+  import { onChange } from '#lib/svelte/svelte'
   import { setIntersection } from '#lib/utils'
+  import { screen } from '#lib/components/stores/screen'
+  import Flash from '#lib/components/flash.svelte'
 
   export let section, displayMode, facets, facetsSelectedValues, textFilterUris
 
-  const { label, entities: works } = section
+  const { entities: works } = section
+  let { label, context } = section
+
+  let filteredWorks = works
+  let paginatedWorks = []
+  let flash
+  if (context) {
+    flash = {
+      type: 'warning',
+      message: context
+    }
+  }
 
-  // TODO: display only the first n items, and add more on scroll
-  let displayedWorks = works
   function filterWorks () {
     if (!facetsSelectedValues) return
     let selectedUris = getSelectedUris({ works, facets, facetsSelectedValues })
     if (textFilterUris) selectedUris = setIntersection(selectedUris, textFilterUris)
-    displayedWorks = works.filter(work => selectedUris.has(work.uri))
+    filteredWorks = works.filter(filterSelectedWorks(selectedUris, facetsSelectedValues))
     if (textFilterUris) {
-      displayedWorks = displayedWorks.sort(bySearchMatchScore(textFilterUris))
+      filteredWorks = filteredWorks.sort(bySearchMatchScore(textFilterUris))
     }
   }
 
+  const filterSelectedWorks = (selectedUris, facetsSelectedValues) => work => {
+    const { uri } = work
+    return selectedUris.has(uri) || isSelectedEntityAParent(uri, facetsSelectedValues)
+  }
+
+  const isSelectedEntityAParent = (uri, facetsSelectedValues) => {
+    // ie. if a collection is selected, display the collection in question.
+    // TODO: generalize pattern to series
+    return facetsSelectedValues['wdt:P195'] === uri
+  }
+
   $: onChange(facetsSelectedValues, textFilterUris, filterWorks)
+
+  function onWorksScroll (e) {
+    const { scrollTop, scrollTopMax } = e.currentTarget
+    if (scrollTopMax < 100) return
+    if (scrollTop + 100 > scrollTopMax) displayLimit += 10
+  }
+
+  // Limit needs to be high enough to have enough elements in order to be scrollable
+  // otherwise on:scroll wont be triggered
+  let initialLimit = 25
+  let displayLimit = initialLimit
+
+  let scrollableElement
+
+  async function addMoreWorks () {
+    paginatedWorks = filteredWorks.slice(0, displayLimit)
+    await addImagesToPaginatedWorks()
+  }
+
+  async function addImagesToPaginatedWorks () {
+    paginatedWorks = await addWorksImages(paginatedWorks)
+  }
+
+  function resetView () {
+    if (scrollableElement) scrollableElement.scroll({ top: 0, behavior: 'smooth' })
+    displayLimit = initialLimit
+  }
+
+  $: onChange(displayLimit, addMoreWorks)
+  $: onChange(displayMode, resetView)
+  $: onChange(filteredWorks, resetView, addMoreWorks)
```

Comment:
This triggers 3 calls to `addMoreWorks` on component initialization: calling a debounced version of `addMoreWorks` could be a work around.

Also, `$: onChange(filteredWorks, resetView, addMoreWorks)` is weird: only the last `onChange` argument is called, all the previous arguments are there to be watched for changes; `resetView` being a constant, it has no effect being there. If the desired behavior is to call both functions, you would need to define another function (ex: `onFilteredWorksChange`) that calls those 2.

Reply:
done differently in 16a8e22 with a early debounce